sdk/search/Azure.Search.Documents/tests/Samples/FieldBuilder/FieldBuilder.cs

```diff
@@ -0,0 +1,451 @@
+
+        /// <summary>
+        /// Creates a collection of <see cref="SearchField"/> objects corresponding to
+        /// the properties of the type supplied.
+        /// </summary>
+        /// <param name="modelType">
+        /// The type for which fields will be created, based on its properties.
+        /// </param>
+        /// <param name="namingPolicy">
+        /// <see cref="JsonNamingPolicy"/> to use.
+        /// Contract resolver that the SearchIndexClient will use.
+        /// This ensures that the field names are generated in a way that is
+        /// consistent with the way the model will be serialized.
+        /// </param>
+        /// <returns>A collection of fields.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="modelType"/> or <paramref name="namingPolicy"/> is null.</exception>
+        public static IList<SearchField> BuildForType(Type modelType, JsonNamingPolicy namingPolicy)
+        {
+            if (modelType is null)
+            { throw new ArgumentNullException(nameof(modelType));
+            }
+
+            if (namingPolicy is null)
+            { throw new ArgumentNullException(nameof(namingPolicy));
+            }
+
+            ArgumentException FailOnNonObjectDataType()
+            {
+                string errorMessage =
+                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " +
+                    "class or struct with public properties.";
+
+                throw new ArgumentException(errorMessage, nameof(modelType));
+            }
+
+            if (ObjectInfo.TryGet(modelType, out ObjectInfo info))
+            {
+                if (info.Properties.Length == 0)
+                {
+                    throw FailOnNonObjectDataType();
+                }
+
+                // Use Stack to avoid a dependency on ImmutableStack for now.
+                return BuildForTypeRecursive(modelType, info, namingPolicy, new Stack<Type>(new[] { modelType }));
+            }
+
+            throw FailOnNonObjectDataType();
+        }
+
+        private static IList<SearchField> BuildForTypeRecursive(
+            Type modelType,
+            ObjectInfo info,
+            JsonNamingPolicy namingPolicy,
+            Stack<Type> processedTypes)
+        {
+            SearchField BuildField(PropertyInfo prop)
+            {
+                static bool ShouldIgnore(Attribute attribute) =>
+                    attribute is JsonIgnoreAttribute || attribute is FieldBuilderIgnoreAttribute;
+
+                IList<Attribute> attributes = prop.GetCustomAttributes(true).Cast<Attribute>().ToArray();
+                if (attributes.Any(ShouldIgnore))
+                {
+                    return null;
+                }
+
+                SearchField CreateComplexField(SearchFieldDataType dataType, Type underlyingClrType, ObjectInfo info)
```

Comment:
Ditto for nested nested functions.

Reply:
This is a straight-up port of track 1 meant for migration and to not waste time majorly changing something we're going to rewrite anyway.